### PR TITLE
Add lm_loss/perplexity metrics and fix MoE load balance loss extraction

### PR DESCRIPTION
**Pull request by gagika**
# Description

Adds `lm_loss` and `perplexity` metrics to pre-training, fixes MoE load
balance loss extraction for scanned architectures, and fixes eval with
synthetic data.

**Metric improvements:**
- Rename `aux["total_loss"]` → `aux["xent_sum"]` to clarify it is the
  unnormalized sum of per-token cross-entropies, distinct from the full
  training objective (`loss` includes MoE lb, indexer, MTP auxiliary losses)
- Add `learning/lm_loss` (= xent_sum / total_weights) and
  `learning/perplexity` (= exp(lm_loss)) to TensorBoard and console output
- Add `eval/avg_perplexity` and `eval/avg_z_loss` to cumulative eval metrics
- Log `moe_lb_loss` in console output when `num_experts > 1`

**MoE load balance loss fix:**
The previous implementation looked up MoE lb loss via a fixed set of hard-coded
intermediate paths. This broke for Gemma4 and other models (like Llama4) that use scanned
blocks, where the path includes dynamic layer indices. Replaced with
`collect_intermediates_by_suffix()`, a new `maxtext_utils` helper that matches
leaves by key-path suffix, working correctly across scanned, scannable-block,
and standard layer layouts. Same fix applied to indexer loss extraction.

**Synthetic eval fix:**
fixed synthetic evals work with `eval_interval > 0`.

# Tests

Verified a run on Gemma4-26B:

```
python3 -m maxtext.trainers.pre_train.train src/maxtext/configs/base.yml   model_name=gemma4-26b   base_output_directory=${BASE_OUTPUT_DIRECTORY?}   dataset_type=hf   hf_path=allenai/c4   hf_name=en   hf_eval_split=validation   tokenizer_type=huggingface   tokenizer_path=google/gemma-4-26b-a4b-it   per_device_batch_size=1   run_name=runner_finetune_gemma4_26b_2   steps=100   enable_checkpointing=false   sharding_tolerance=0.03   eval_interval=20   eval_steps=5   load_balance_loss_weight=0.1
```

output: https://paste.googleplex.com/5891139803676672

DPO test:

```
 python3 -m maxtext.trainers.pre_train.train src/maxtext/configs/post_train/dpo.yml   run_name=dpo_smoke   base_emb_dim=256   base_num_query_heads=4   base_num_kv_heads=4   base_mlp_dim=512   base_num_decoder_layers=4   head_dim=64   per_device_batch_size=1   max_target_length=128   steps=10   eval_interval=5   eval_steps=2   enable_checkpointing=false
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/common/metric_logger.py [hunk 1]
```diff
@@ -381,6 +392,9 @@ def record_eval_metrics(self, step, metrics=None, eval_step_count=None):
       self.cumulative_eval_metrics["scalar"]["eval/avg_mtp_acceptance_rate_percent"] = (
           self.cumulative_eval_metrics["scalar"]["eval/mtp_acceptance_rate_percent"] / eval_step_count
       )
+      self.cumulative_eval_metrics["scalar"]["eval/avg_z_loss"] = (
```
[the remark this answers is not in the document]
Comment (gagika) [reply src/maxtext/common/metric_logger.py:398]: acknowledged, same pattern as avg_moe_lb_loss/avg_mtp_loss; acceptable for a small regularization term."

File: src/maxtext/trainers/pre_train/train.py [hunk 2]
```diff
@@ -236,15 +237,7 @@ def loss_fn(model, config, data, dropout_rng, params, is_train=True):
...
-    indexer_losses = []
-    # Extract 'indexer_loss' from model intermediates.
-    # We check for paths ending in ('self_attention', 'indexer_loss').
-    # This handles varying paths caused by different layer names.
-    for path, val in jax.tree_util.tree_leaves_with_path(intermediate_outputs):
-      path_keys = tuple(k.key for k in path if hasattr(k, "key"))
-      if path_keys[-2:] == ("self_attention", "indexer_loss"):
-        indexer_losses.append(jnp.ravel(val))
-
+    indexer_losses = maxtext_utils.collect_intermediates_by_suffix(intermediate_outputs, "self_attention", "indexer_loss")
```
Comment (RissyRan) [inline src/maxtext/trainers/pre_train/train.py:239]: We just claimed the support, and could you help have a sanity check? - [cmd in Continued pre-training for V3.2 Sparse Attention](https://github.com/AI-Hypercomputer/maxtext/blob/4909a0afb1ca63cec03d83fdaefadc01b2982bc8/tests/end_to_end/tpu/deepseek/Run_DeepSeek.md#L4). Either dense/sparse training stage should be good.

**Review by RissyRan — COMMENTED** [review]
Thanks Gagik! For the balance loss, you may need to set it for dropping, which needs `capacity_factor=1 sparse_matmul=False`. After 50-100 training steps, you should be able to see something like [this](https://screenshot.googleplex.com/AFEHtmpoT3gT8oB). Could you help have a sanity check? I see the loss is stable in your logs.

File: src/maxtext/trainers/pre_train/train.py (hunk 2 above)
Comment (gagika) [reply src/maxtext/trainers/pre_train/train.py:239]: tested that `learning/indexer_loss` is working:

python3 -m maxtext.trainers.pre_train.train src/maxtext/configs/base.yml   base_output_directory=$BASE_OUTPUT_DIRECTORY   run_name=indexer_sparse_1   model_name=deepseek3.2-671b   tokenizer_type=huggingface tokenizer_path=deepseek-ai/DeepSeek-V3.2   per_device_batch_size=1 enable_checkpointing=false async_checkpointing=false   steps=5 max_target_length=256 attention=dot_product   dtype=bfloat16 weight_dtype=bfloat16 megablox=True sparse_matmul=True   dataset_type=synthetic indexer_loss_scaling_factor=0.01   indexer_sparse_training=True gcs_metrics=True override_model_config=True   first_num_dense_layers=1 base_num_decoder_layers=2 base_moe_mlp_dim=512   scan_layers=False load_balance_loss_weight=0.1  2>&1 indexer_topk=128 max_target_length=256

https://screenshot.googleplex.com/4UA3EvD3uYVCwdt

added a check that  indexer_topk <= max_target_length

**Comment by gagika** [comment]
> Thanks Gagik! For the balance loss, you may need to set it for dropping, which needs `capacity_factor=1 sparse_matmul=False`. After 50-100 training steps, you should be able to see something like [this](https://screenshot.googleplex.com/AFEHtmpoT3gT8oB). Could you help have a sanity check? I see the loss is stable in your logs.

done:

https://screenshot.googleplex.com/3FcUcvjd4YooesH

File: src/maxtext/configs/types.py [hunk 3]
```diff
@@ -2530,6 +2530,12 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
         raise NotImplementedError(
             "Sparse indexer is only supported dot_product attention or flash attention with tokamax splash."
         )
+      if self.indexer_sparse_training and self.indexer_topk >= self.max_target_length:
```
Comment (RissyRan) [inline src/maxtext/configs/types.py]: This only applies to sparse training, and I think we could include both sparse & dense Indexer training.

https://github.com/AI-Hypercomputer/maxtext/blob/51c7f2b8e5a53c3cf0802691d59e0ff1684483ab/src/maxtext/layers/attention_mla.py#L1197

This is the condition when loss is enabled. I think properly we could have something like:

`use_indexer and  self.config.indexer_loss_scaling_factor > 0.0 and self.indexer_topk >= self.max_target_length`

Comment (gagika) [reply src/maxtext/configs/types.py]: done, please have a look

**Review by RissyRan — APPROVED** [review]

File: src/maxtext/common/metric_logger.py [hunk 4]
```diff
@@ -190,6 +201,7 @@ def _log_eval_metrics(self, metrics, step):
     log_parts = [
         f"eval metrics after step: {step}",
         f"loss={scalars['eval/avg_loss']:.3f}",
+        f"perplexity={scalars['eval/avg_perplexity']:.3f}",
```
Comment (JamesDeng42) [inline src/maxtext/common/metric_logger.py:204]: use scalars.get('eval/avg_perplexity', 0.0) to avoid crash if not existed.

**Review by JamesDeng42 — APPROVED** [review]
Generally LGTM , left one comment.

File: src/maxtext/common/metric_logger.py (hunk 4 above)
Comment (gagika) [reply src/maxtext/common/metric_logger.py:204]: The `eval/avg_perplexity` is always populated alongside `eval/avg_loss`, so the key will always exist when _log_eval_metrics is called. 
I think there are a few tasks that perplexity might not make sense, e.g. index trainer, but we still populate it with 0, keeping like this will cause crash, which is probably better for many situation that there is a bug.
let me know what you think?
